moz_sql_parser/formatting.py

```diff
@@ -0,0 +1,242 @@
+                return self.op(json)
+        if isinstance(json, string_types):
+            return escape(json, self.ansi_quotes, self.should_quote)
+        
+        return text_type(json)
+
+    def delimited_list(self, json):
+        return ', '.join(self.dispatch(element) for element in json)
+
+    def value(self, json):
+        parts = [self.dispatch(json['value'])]
+        if 'name' in json:
+            parts.extend(['AS', self.dispatch(json['name'])])
+        return ' '.join(parts)
+
+    def op(self, json):
+        if 'on' in json:
+            return self._on(json)
+
+        if len(json) > 1:
+            raise Exception('Operators should have only one key!')
+        key, value = list(json.items())[0]
+
+        # check if the attribute exists, and call the corresponding method;
+        # note that we disallow keys that start with `_` to avoid giving access
+        # to magic methods
+        attr = '_{0}'.format(key)
+        if hasattr(self, attr) and not key.startswith('_'):
+            method = getattr(self, attr)
+            return method(value)
+
+        return '{0}({1})'.format(key.upper(), value)
+
+    def _exists(self, value):
+        return '{0} IS NOT NULL'.format(self.dispatch(value))
+
+    def _missing(self, value):
+        return '{0} IS NULL'.format(self.dispatch(value))
+
+    def _like(self, pair):
+        return '{0} LIKE {1}'.format(self.dispatch(pair[0]), self.dispatch(pair[1]))
+
+    def _is(self, pair):
+        return '{0} IS {1}'.format(self.dispatch(pair[0]), self.dispatch(pair[1]))
+
+    def _in(self, json):
+        valid = self.dispatch(json[1])
+        # `(10, 11, 12)` does not get parsed as literal, so it's formatted as
+        # `10, 11, 12`. This fixes it.
+        if not valid.startswith('('):
+            valid = '({0})'.format(valid)
+
+        return '{0} IN {1}'.format(json[0], valid)
+
+    def _case(self, checks):
+        parts = ['CASE']
+        for check in checks:
+            if isinstance(check, dict):
+                parts.extend(['WHEN', self.dispatch(check['when'])])
+                parts.extend(['THEN', self.dispatch(check['then'])])
+            else:
+                parts.extend(['ELSE', self.dispatch(check)])
+        parts.append('END')
+        return ' '.join(parts)
+
+    def _literal(self, json):
+        if isinstance(json, list):
+            return '({0})'.format(', '.join(self._literal(v) for v in json))
+        elif isinstance(json, string_types):
+            return "'{0}'".format(json.replace("'", "''"))
+        else:
+            return str(json)
+
+    def _on(self, json):
+        return 'JOIN {0} ON {1}'.format(
+            self.dispatch(json['join']), self.dispatch(json['on']))
+
+    def union(self, json):
+        return ' UNION '.join(self.query(query) for query in json)
+
+    def query(self, json):
+        parts = []
+        for clause in self.clauses:
+            method = getattr(self, clause, None)
+            if method:
```

Comment:
I think the method always exists

Reply:
You're right, this was leftover from when I was developing the code and hadn't implemented all the methods yet. I'll do a small PR addressing your comments on this one.